[qa] Add a timeout parameter to sync_blocks

sickpig opened the pull request:
Default value is 60 seconds, could be changed by passing a "timeout" (named only parameter) when calling the method.

sickpig committed:
[qa] Add a timeout parameter to sync_blocks

Default value is 60 seconds, could be changed by passing
a "timeout" (named only parameter) when calling the method.

sickpig commented:
Travis running at
https://travis-ci.org/github/BitcoinUnlimited/BitcoinUnlimited/jobs/666889902

gandrewstone commented:
I think you'd want to raise an exception if the timeout has been exceeded, since that indicates a failed test (once the code falls thru the while loop).  For the few cases where a timeout is OK, the caller can catch the exception.

Change some test's initial sync_blocks to call this fn with a very short timeout and catch the exception to test its behavior.

sickpig commented:
> I think you'd want to raise an exception if the timeout has been exceeded, since that indicates a failed test (once the code falls thru the while loop). For the few cases where a timeout is OK, the caller can catch the exception.

Is there any test for which a timing out `sync_blocks()` is ok? 

As I see it all tests that put `sync_blocks()` in place expect that the sync actually happens for the test to complete successfully. if it weren't not the case `sync_blocks()` call would have been superfluous.

I don't have a strong opinion about raising an error when the timeout period expired, but I guess that this wouldn't do any difference because I would expect the test to fail anyway in a subsequent assert.

sickpig committed:
Raise an exception if the timeout kicks in for sync_blocks()

sickpig commented:
@gandrewstone 
added the code to raise an exception if timeout kicks in, tested in a bunch of units to see if it works by decreasing the default of 60 seconds to 1 second. 
worked all the time

ptschip commented on qa/rpc-tests/test_framework/util.py:
```diff
@@ -279,15 +279,16 @@ def is_connected(gdict, vertices_encountered = None, start_vertex=None):
         counts = [ x.getblockcount() for x in rpc_connections ]
         if counts == [ counts[0] ]*len(counts):
-            break
+            return
```
shouldn't you take out the return and leave it as "break"?

sickpig replied:
mmm the break will bring the code of out of the while loop and then we are going to raise the timeout exception  even in the case the timeout won't get reached, no?

ptschip replied:
duh, yes right...

ptschip commented:
One last thing, to be consistent , shouldn't we also add the same functionality to the function "sync_blocks_to()" ?

sickpig commented:
> One last thing, to be consistent , shouldn't we also add the same functionality to the function "sync_blocks_to()" ?

you have a point, looking into it.

sickpig committed:
Add a timeout to sync_blocks_to()

The sam,e way that has been done with sync_blocks() the default timeout
value is 60 seconds.

sickpig commented:
@ptschip added the same timeout machienery to `sync_blocks_to()`, tested with `ctor.py` which is the only tests the use that method.
